packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js

```diff
@@ -170,24 +179,124 @@ export default class SidebarAdapter {
 			'',
 			settingArgs
 		);
-		setting.set( {} );
+		setting.set( widget.instance );
 
-		const widgetIds = [ ...this.setting.get() ];
 		const widgetId = settingIdToWidgetId( settingId );
-		widgetIds.splice( index, 0, widgetId );
-		this.setting.set( widgetIds );
 
 		return widgetId;
 	}
 
-	updateWidget( widget ) {
+	_updateWidget( widget ) {
+		const prevWidget = this.getWidget( widget.id );
+
+		// Bail out update.
+		if ( prevWidget === widget ) {
+			return widget.id;
+		}
+
 		const settingId = widgetIdToSettingId( widget.id );
 		this.allSettings( settingId ).set( widget.instance );
 		// TODO: what about the other stuff?
+
+		return widget.id;
 	}
 
-	removeWidget( widgetId ) {
-		const widgetIds = this.setting.get();
-		this.setting.set( without( widgetIds, widgetId ) );
+	getWidget( widgetId ) {
+		if ( ! widgetId ) {
+			return null;
+		}
+
+		const { idBase, number } = parseWidgetId( widgetId );
+		const settingId = widgetIdToSettingId( widgetId );
+		const setting = this.allSettings( settingId );
+
+		if ( ! setting ) {
+			return null;
+		}
+
+		const instance = setting.get();
+
+		if ( this.widgetsMap.has( instance ) ) {
+			return this.widgetsMap.get( instance );
+		}
+
+		const widget = {
+			id: widgetId,
+			idBase,
+			number,
+			instance,
+		};
+
+		this.widgetsMap.set( instance, widget );
+
+		return widget;
+	}
+
+	_internalUpdateWidgets( nextWidgets ) {
+		this.locked = true;
+
+		const addedWidgetIds = [];
+
+		const nextWidgetIds = nextWidgets.map( ( nextWidget ) => {
+			if ( nextWidget.id && this.getWidget( nextWidget.id ) ) {
+				addedWidgetIds.push( null );
+
+				return this._updateWidget( nextWidget );
+			}
+
+			const widgetId = this._createWidget( nextWidget );
+
+			addedWidgetIds.push( widgetId );
+
+			return widgetId;
+		} );
+
+		// TODO: We should in theory also handle delete widgets here too.
+
+		this.setting.set( nextWidgetIds );
+
+		this.locked = false;
+
+		return addedWidgetIds;
+	}
+
+	setWidgets( nextWidgets ) {
+		const addedWidgetIds = this._internalUpdateWidgets( nextWidgets );
+
+		this._pushHistory();
+
+		return addedWidgetIds;
+	}
+
+	canUndo() {
+		return this.historyIndex > 0;
+	}
+	canRedo() {
+		return this.historyIndex < this.history.length - 1;
+	}
+	_seek( historyIndex ) {
+		const currentWidgets = this.getWidgets();
+
+		this.historyIndex = historyIndex;
+
+		const widgets = this.history[ this.historyIndex ];
+
+		this._internalUpdateWidgets( widgets );
+
+		this._emit( currentWidgets, this.getWidgets() );
+	}
+	undo() {
+		if ( ! this.canUndo() ) {
+			return;
+		}
+
+		this._seek( this.historyIndex - 1 );
+	}
+	redo() {
+		if ( ! this.canRedo() ) {
+			return;
+		}
+
+		this._seek( this.historyIndex + 1 );
```

Comment:
These are currently unused, but paving the way for the undo/redo feature in the future.